core/rid_handle.cpp

```diff
@@ -0,0 +1,303 @@
+		}
+#endif
+	}
+
+#ifdef RID_HANDLE_ALLOCATION_TRACKING_ENABLED
+	for (uint32_t n = 0; n < _leaks.size(); n++) {
+		const Leak &leak = _leaks[n];
+
+		const char *tn = "RID_Owner unknown";
+		if (_owner_names.size()) {
+			tn = _owner_names[leak.owner_name_id];
+		}
+
+		const char *fn = "Filename unknown";
+		if (leak.filename) {
+			fn = leak.filename;
+		}
+
+		_err_print_error(tn, fn, leak.line_number, itos(leak.num_objects_leaked) + " RID objects leaked");
+	}
+#endif
+
+	_shutdown = true;
+}
+
+void RID_Database::handle_make_rid(RID &r_rid, RID_Data *p_data, RID_OwnerBase *p_owner) {
+	ERR_FAIL_COND_MSG(_shutdown, "RID_Database make_rid use after shutdown.");
+	ERR_FAIL_COND_MSG(!p_data, "RID_Database make_rid, data is empty.");
+	bool data_was_empty = true;
+
+	_mutex.lock();
+	PoolElement *pe = _pool.request(r_rid._id);
+	data_was_empty = !pe->data;
+
+	pe->data = p_data;
+	p_data->_owner = p_owner;
+	pe->revision = pe->revision + 1;
+	r_rid._revision = pe->revision;
+
+#ifdef RID_HANDLE_ALLOCATION_TRACKING_ENABLED
+	pe->line_number = 0;
+	pe->filename = nullptr;
+#endif
+
+	_mutex.unlock();
+
+	ERR_FAIL_COND_MSG(!data_was_empty, "RID_Database make_rid, previous data was not empty.");
+}
+
+RID_Data *RID_Database::handle_get(const RID &p_rid) {
+	RID_Data *data = handle_get_or_null(p_rid);
+#ifdef RID_HANDLE_FLAG_NULL_GETS
+	ERR_FAIL_COND_V_MSG(!data, nullptr, "RID_Database get is NULL");
+#endif
+	return data;
+}
+
+RID_Data *RID_Database::handle_getptr(const RID &p_rid) {
+	RID_Data *data = handle_get_or_null(p_rid);
+#ifdef RID_HANDLE_FLAG_NULL_GETS
+	ERR_FAIL_COND_V_MSG(!data, nullptr, "RID_Database getptr is NULL");
+#endif
+	return data;
+}
+
+// Note, no locks used in the getters.
+// Godot 4.x does use locks in the getters, but it is arguably overkill because even though
+// the pointer returned will be correct (i.e. it has not been replaced during this call),
+// it can be invalidated during the client code use. (There may also be an internal reason why
+// locks are needed in 4.x, as the database is different.)
+// An example of a "safer" way to do this kind of thing is object level locking,
+// (but that has complications of its own), or atomic object changes.
+
+RID_Data *RID_Database::handle_get_or_null(const RID &p_rid) {
+	if (p_rid.is_valid()) {
+		ERR_FAIL_COND_V_MSG(_shutdown, nullptr, "RID_Database get_or_null after shutdown.");
+
+		// The if statement is to allow breakpointing without a recompile.
+		if (p_rid._id >= _pool.pool_reserved_size()) {
+			ERR_FAIL_COND_V_MSG(p_rid._id >= _pool.pool_reserved_size(), nullptr, "RID_Database get_or_null, RID id was outside pool size.");
+		}
+
+		const PoolElement &pe = _pool[p_rid._id];
+		if (pe.revision != p_rid._revision) {
+			print_verbose("RID revision incorrect : " + _rid_to_string(p_rid, pe));
+			ERR_FAIL_COND_V_MSG(pe.revision != p_rid._revision, nullptr, "RID_Database get_or_null, revision is incorrect, object possibly freed before use.");
+		}
+
```

Comment:
Note to myself: In retrospect, I probably should add a lock here if we do decide to roll out further.

I suspect there is a potential race condition from getting `PoolElement` and reading at the exact moment a pool resize is happening from another thread (this should be incredibly rare, and may not even happen in practice). This is kind of a pain, as it would be nice to have this performance sensitive function operate without a lock, especially for such an unlikely situation. Maybe there is a way around this. Anyway not a significant problem unless we do decide to make regular builds with handles.